Implemented non-named version of unflatten

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42613 Implemented torch::nn::Unflatten in libtorch
* **#42563 Implemented non-named version of unflatten**

Moved logic for non-named unflatten from python nn module to aten/native to be reused by the nn module later. Fixed some inconsistencies with doc and code logic.

Differential Revision: [D23030301](https://our.internmc.facebook.com/intern/diff/D23030301)

heitorschueroff committed:
Implemented non-named version of unflatten

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented non-named version of unflatten"




[ghstack-poisoned]

zou3519 commented on torch/tensor.py:
```diff
@@ -647,25 +647,36 @@ def align_to(self, *names):
 
         Arguments:
-            namedshape: (iterable of ``(name, size)`` tuples).
+            dim (Union[str, int]): Dimension to unflatten
+            sizes (Union[Iterable[Tuple[str, int]], torch.Size]): New shape of the unflattened dimension
```
Nit: prefer `int` to come first in dim, and `torch.Size` to come first in `sizes`. Most people won't be using this for the named tensor functionality.

zou3519 commented on test/test_torch.py:
```diff
@@ -2463,6 +2463,11 @@ def test_flatten(self):
                     src.flatten(2, 0)
 
+        def test_unflatten(self):
+            self.assertEqual(torch.tensor([]).unflatten(0, (0, 1, 0)), torch.empty(0, 1, 0))
+            self.assertEqual(torch.tensor([1]).unflatten(0, (1, 1)), torch.tensor([[1]]))
+            self.assertEqual(torch.tensor([[1, 2, 3, 4]]).unflatten(1, (2, 2)), torch.tensor([[[1, 2], [3, 4]]]))
+
```
Could we also test the following:
- pass in a torch.Size for `sizes` (since we advertise we support that)
- pass in a List[int] for `sizes` (all of our functions that take torch.Size also support List[int])
- pass in an empty list for sizes (what happens in this case?)
- check the error message when passing in an invalid size (e.g., `torch.randn(3).unflatten([1, 1])`)
- call unflatten on a scalar tensor (`torch.tensor(1).unflatten(0, ??)`)

To test error messages one can use assertRaisesRegex. Example: https://github.com/pytorch/pytorch/blob/b85216887ba76dc7e1e0035c77ebe358ac48f90d/test/test_torch.py#L123-L124

glaringlee reviewed (commented):
Other than reviews from @zou3519 
Two things want to confirm here @zou3519 @heitorschueroff :
1. This is not really a bc breaking change, but it breaks bc CI test, I think this is one time breaking after checking the check_backward_compatibility.py, but need to confirm.
2. I think unflatten should be added into _overrides.py here like flatten:
https://github.com/pytorch/pytorch/blob/206db5c1272a536fc668a1b5822a7de2c4fc5934/torch/_overrides.py#L314

zou3519 commented:
> Aug 04 23:06:48 Broken ops: [
Aug 04 23:06:48 	aten::unflatten.Dimname(Tensor self, str dim, int[] sizes, str[] names) -> (Tensor)
Aug 04 23:06:48 	aten::unflatten.int(Tensor self, int dim, int[] sizes, str[] names) -> (Tensor)
Aug 04 23:06:48 ]

I'm pretty confident no one has serialized the above operators because they literally don't work in TorchScript (we've banned named tensors in TorchScript). So I wouldn't worry about the BC-breakage.

I remember there was a way to "approve" a change to the backwards compatibility test so the PR doesn't break anything when it gets merged. cc @houseroad -- how does one do that again?

glaringlee commented:
> Other than reviews from @zou3519
> Two things want to confirm here @zou3519 @heitorschueroff :
> 
> 1. This is not really a bc breaking change, but it breaks bc CI test, I think this is one time breaking after checking the check_backward_compatibility.py, but need to confirm.
> 2. I think unflatten should be added into _overrides.py here like flatten:
>    https://github.com/pytorch/pytorch/blob/206db5c1272a536fc668a1b5822a7de2c4fc5934/torch/_overrides.py#L314

**Update** No need to add unflatten into _overrides.py, since we don't support torch.unflatten.

heitorschueroff committed:
Update on "Implemented non-named version of unflatten"


Moved logic for non-named unflatten from python nn module to aten/native to be reused by the nn module later. Fixed some inconsistencies with doc and code logic.

[ghstack-poisoned]

glaringlee commented:
> > Aug 04 23:06:48 Broken ops: [
> > Aug 04 23:06:48 	aten::unflatten.Dimname(Tensor self, str dim, int[] sizes, str[] names) -> (Tensor)
> > Aug 04 23:06:48 	aten::unflatten.int(Tensor self, int dim, int[] sizes, str[] names) -> (Tensor)
> > Aug 04 23:06:48 ]
> 
> I'm pretty confident no one has serialized the above operators because they literally don't work in TorchScript (we've banned named tensors in TorchScript). So I wouldn't worry about the BC-breakage.
> 
> I remember there was a way to "approve" a change to the backwards compatibility test so the PR doesn't break anything when it gets merged. cc @houseroad -- how does one do that again?

Per @houseroad , please add unflatten function here https://fburl.com/lsb7ah1y
The date should be 1 week after expected landing time.

heitorschueroff commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1531,6 +1531,44 @@ Tensor flatten(const Tensor& self, DimnameList dims, Dimname out_dim) {
+  Tensor result;
+  {
+    NoNamesGuard guard;
+    result = self.reshape(shape);
```
flatten uses reshape so I wanted to make it consistent with unflatten. Maybe we should change flatten to use view instead?

zou3519 replied:
The reason flatten uses `reshape` is because it is not always possible to flatten dimensions together. Take for instance:
```
x = torch.randn(2, 3).t()
x.view(6) # fails due to discontiguity
```
Now, we can always unflatten a single dimension into multiple dimensions by performing a view. Because of that, it doesn't matter if we call `reshape` or `view` -- the returned tensor will always be a view (reshape has two paths: if reshape can perform a view, then it will perform the view. Otherwise, it will perform a copy). 

Since `reshape` will always call `view` (and never perform a copy), we should make the semantics clear and do a `view` here.

heitorschueroff replied:
I see, thank you for clarifying.

heitorschueroff commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1315,6 +1309,12 @@
 - func: flatten.DimnameList(Tensor(a) self, Dimname[] dims, Dimname out_dim) -> Tensor(a)
   variants: function, method
 
+- func: unflatten.int(Tensor(a) self, int dim, int[] sizes, Dimname[]? names=None) -> Tensor(a)
+  variants: method
+
```
Like I mentioned in the comment below, flatten uses reshape while unflatten originally used view. I think we should make them both consistent and then depending on which one I'll update tensor_view.rst to include both functions.

heitorschueroff commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1531,6 +1531,44 @@ Tensor flatten(const Tensor& self, DimnameList dims, Dimname out_dim) {
+    result = self.reshape(shape);
+  }
+
+  if (names) {
+    auto outnames = self.names().vec();
+    outnames.erase(outnames.begin() + dim);
+    outnames.insert(outnames.begin() + dim, names->begin(), names->end());
+    at::internal_set_names_inplace(result, outnames);
+  }
```
I think asking user to provide names is better but wouldn't it be BC-breaking?

zou3519 replied:
We've never supported calling `x.unflatten(1, [2, 5])` on a named tensor (this PR adds that functionality) so I don't think this is BC-breaking. e.g. this is what happens on master:
```
>>> x = torch.randn(2, 10, names=('N', 'D'))
>>> x.unflatten(1, [2, 5])
TypeError: zip argument #1 must support iteration
```
Is there a case you were thinking of?

heitorschueroff committed:
Update on "Implemented non-named version of unflatten"


Moved logic for non-named unflatten from python nn module to aten/native to be reused by the nn module later. Fixed some inconsistencies with doc and code logic.

[ghstack-poisoned]

glaringlee commented:
@heitorschueroff 
All looks good to me now. One more thing, please fix flake8-py3 lint error

heitorschueroff committed:
Update on "Implemented non-named version of unflatten"


Moved logic for non-named unflatten from python nn module to aten/native to be reused by the nn module later. Fixed some inconsistencies with doc and code logic.

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented non-named version of unflatten"


Moved logic for non-named unflatten from python nn module to aten/native to be reused by the nn module later. Fixed some inconsistencies with doc and code logic.

[ghstack-poisoned]

glaringlee reviewed (approved):
This LGTM now

zou3519 commented on torch/tensor.py:
```diff
@@ -807,31 +807,46 @@ def align_to(self, *names):
+        r"""Expands the dimension :attr:`dim` of the :attr:`self` tensor over multiple dimensions
+        of sizes given by :attr:`sizes`.
+
+        * :attr:`sizes` is the new shape of the unflattened dimension and it can be a `Iterable[int]` as well
```
Can it actually be an `Iterable[int]`? iirc our PythonArgParser doesn't handle Iterable[int]. If unflatten can't take in an iterable of ints, I would just say tuple of ints here.

Here's how to create an iterable of ints: `iter([2, 3, 4])`

glaringlee replied:
@zou3519  Ah, this is something I don't know......

heitorschueroff replied:
Hmm, so we only support tuple, list and torch.Size is that right?

zou3519 commented on torch/tensor.py:
```diff
@@ -807,31 +807,46 @@ def align_to(self, *names):
             return handle_torch_function(Tensor.unflatten, relevant_args, self, dim, namedshape)
-        names, sizes = unzip_namedshape(namedshape)
+
+        names = None
+        if isinstance(dim, str) or self.has_names():
+            try:
+                names, sizes = unzip_namedshape(sizes)
+            except TypeError:
+                raise TypeError("unflatten: Expected sizes to be of type Iterable[Tuple[str, int]] but got " + type(sizes).__name__)
         return super(Tensor, self).unflatten(dim, sizes, names)
```
Sorry, I wasn't clear here with my previous comment.

Before this PR, we supported the following:
- unflatten(Tensor, int, namedshape)
- unflatten(NamedTensor, int, namedshape)
- unflatten(Tensor, str, namedshape)
- unflatten(NamedTensor, str, namedshape)

It looks like we've dropped support for the case of
- unflatten(Tensor, int, namedshape)

Ideally, we'd keep BC. My suggestion here is the following:
- check if `sizes` is a namedshape.
    - if it is a namedshape, then parse it into `sizes` and `names`
    - if it's not a namedshape, then pass in `sizes` and `names = None`
- if `sizes` is not a namedshape and the input is a named tensor, then we raise an error. Previously we had required `sizes` to always be a namedshape. Now we are relaxing that constraint. Since we are adding new behavior here, this error isn't BC-breaking

zou3519 replied:
NB: It's not the biggest problem in the world if we break BC here. I don't think many users were using the `unflatten(Tensor, int, namedshape)` case, and we advertise named tensors as "experimental and subject to change without warning".

glaringlee reviewed (commented):
@zou3519 
Agree unflatten(Tensor, int, namedshape) should be handled.
Just curious, how we handle unflatten(Tensor, str, namedshape)? if input tensor is not named, will user pass `str`? 

@heitorschueroff I think we should add back unflatten(Tensor, int, namedshape)` support

zou3519 commented:
> How about this one? unflatten(Tensor, str, namedshape), if input tensor is not named, will user pass `str`?

On master that errors out, so I'd expect it to give a nice error after this PR

glaringlee reviewed (changes requested):
per above comments

heitorschueroff committed:
Update on "Implemented non-named version of unflatten"


Moved logic for non-named unflatten from python nn module to aten/native to be reused by the nn module later. Fixed some inconsistencies with doc and code logic.

Differential Revision: [D23030301](https://our.internmc.facebook.com/intern/diff/D23030301)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented non-named version of unflatten"


Moved logic for non-named unflatten from python nn module to aten/native to be reused by the nn module later. Fixed some inconsistencies with doc and code logic.

Differential Revision: [D23030301](https://our.internmc.facebook.com/intern/diff/D23030301)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented non-named version of unflatten"


Moved logic for non-named unflatten from python nn module to aten/native to be reused by the nn module later. Fixed some inconsistencies with doc and code logic.

Differential Revision: [D23030301](https://our.internmc.facebook.com/intern/diff/D23030301)

[ghstack-poisoned]

zou3519 commented on test/test_namedtensor.py:
```diff
@@ -1062,12 +1062,38 @@ def test_flatten(self):
+            torch.ones(2, 2, 2, names=('A', 'B1', 'B2'))))
+
+        # test invalid args: namedtensor, str, sizes
+        self.assertRaises(TypeError, lambda: torch.tensor([1], names=('A',)).unflatten('A', (1, 1)))
```
nit: it's nice to use self.assertRaisesRegex to ensure that the error message is human readable

zou3519 commented on test/test_torch.py:
```diff
@@ -2463,6 +2463,30 @@ def test_flatten(self):
+            self.assertEqual(torch.ones(2, 10).unflatten(1, (5, 2)), torch.ones(2, 5, 2))
+
+            # test invalid args: tensor, str, sizes
+            self.assertRaises(TypeError, lambda: torch.tensor([1]).unflatten('A', (1, 1)))
```
nit: I prefer to use self.assertRaisesRegex to ensure that the error message is human readable

zou3519 commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1531,6 +1531,48 @@ Tensor flatten(const Tensor& self, DimnameList dims, Dimname out_dim) {
+      dim, " (", self.size(dim), ") in the input tensor");
+  }
+
+  auto shape = self.sizes().vec();
```
nit: Using a DimVector gives better performance because it avoids dynamic allocations on the heap when the tensor has <= 5 dimensions: https://github.com/pytorch/pytorch/blob/4afbf3973706f349a61447649e34447356fb03cc/aten/src/ATen/core/DimVector.h#L11

zou3519 replied:
IIRC the usage would be: `DimVector shape(self.sizes().begin(), self.sizes().end());`

zou3519 reviewed (approved):
LGTM, thanks for addressing all the comments!

I had some minor nits left, feel free to address (or not)

heitorschueroff committed:
Update on "Implemented non-named version of unflatten"


Moved logic for non-named unflatten from python nn module to aten/native to be reused by the nn module later. Fixed some inconsistencies with doc and code logic.

Differential Revision: [D23030301](https://our.internmc.facebook.com/intern/diff/D23030301)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented non-named version of unflatten"


Moved logic for non-named unflatten from python nn module to aten/native to be reused by the nn module later. Fixed some inconsistencies with doc and code logic.

Differential Revision: [D23030301](https://our.internmc.facebook.com/intern/diff/D23030301)

[ghstack-poisoned]
